### ISSUE TITLE
F821: false positive in nested func

### ISSUE THREAD

**Issue by SoundDesignerToBe** [astral-sh/ruff#18610]
### Summary

I just started using ruff as a pre-commit checker/formatter/linter and ran into an issue, where
"undefined" variable is reported as an error, but the code runs because it is a nested function. 

This is probably related to other issues w.r.t. F821 but their titles appeared to reference other scenarios. 
Sorry if I produced a duplicate report here. 

```python
# test_my_code.py
import pytest

def test_my_code():
    application = Application(...)

    def check_window(app):
        global test_success
        windows = [
            win
            for win in Gtk.Window.list_toplevels()
            if hasattr(win, "application") and win.application == application
        ]
        ...

    def decorator2(method):
        global test_success

        def wrapper(*args, **kwargs):
            method(*args, **kwargs)
            GLib.idle_add(check_window, args[0])

        return wrapper
```

```shell
tests/...:67:67: F821 Undefined name `application`
   |
65 |             win
66 |             for win in Gtk.Window.list_toplevels()
67 |             if hasattr(win, "application") and win.application == application
   |                                                                   ^^^^^^^^^^^ F821
68 |         ]
69 |         try:
   |

Found 1 error.

```

This is my pre-commit hook
```yaml
repos:
  - repo: https://github.com/astral-sh/ruff-pre-commit
    rev: v0.11.12
    hooks:       
      - id: ruff  # e.g. missing imports, non-declared variables/commands, ...
        name: ruff-unresolved-references
        args: [--select=F821]
        files: ^(src/|tests/)
```

### Version

v0.11.13

**Comment by ntBre** [comment]
I'm having some trouble reproducing this locally or in the [playground](https://play.ruff.rs/9bec307f-ef60-4735-8ea1-c1a8edcb8464). Is there possibly a `del` statement or other important code omitted from the example?
